core/src/main/java/greencity/validators/AddressValidator.java

```diff
@@ -0,0 +1,84 @@
+package greencity.validators;
+
+import com.google.maps.model.AddressComponent;
+import com.google.maps.model.AddressComponentType;
+import com.google.maps.model.GeocodingResult;
+import com.google.maps.model.LatLng;
+import greencity.annotations.ValidAddress;
+import greencity.dto.CreateAddressRequestDto;
+import greencity.dto.google.AddressResponseFromGoogleAPI;
+import greencity.dto.location.CoordinatesDto;
+import greencity.exceptions.address.InvalidAddressException;
+import greencity.exceptions.api.GoogleApiException;
+import greencity.service.google.GoogleApiService;
+import jakarta.validation.ConstraintValidator;
+import jakarta.validation.ConstraintValidatorContext;
+import java.util.Arrays;
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+public class AddressValidator implements ConstraintValidator<ValidAddress, CreateAddressRequestDto> {
+    private final GoogleApiService googleApiService;
+    private static final double DELTA = 0.007;
+    private static final int LANGUAGE_CODE_FOR_UA = 0;
+
+    @Override
+    public boolean isValid(CreateAddressRequestDto createAddressRequestDto, ConstraintValidatorContext context) {
+        String placeId = createAddressRequestDto.getPlaceId();
+        CoordinatesDto coordinates = createAddressRequestDto.getCoordinates();
+        LatLng latLng = new LatLng(coordinates.getLatitude(), coordinates.getLongitude());
+
+        GeocodingResult geoResult;
+        AddressResponseFromGoogleAPI resultFromCoordinates;
+
+        try {
+            geoResult = googleApiService.getResultFromGeoCode(placeId, LANGUAGE_CODE_FOR_UA);
```

Comment:
Please add changing to not call `googleApi` every timefor the same `placeId`

Reply:
i added cache to this method:
``` 
@Cacheable(value = "geoCodeCache", key = "#placeId + '-' + #langCode")
    public GeocodingResult getResultFromGeoCode(String placeId, Integer langCode) {